wkcuber/wkcuber/tile_cubing.py

```diff
@@ -185,89 +182,92 @@ def find_file_with_dimensions(
-                                    np.zeros(
-                                        tile_size + (1,),
-                                        dtype=target_wkw_info.header.voxel_type,
-                                    )
-                                )
-                        buffer = prepare_slices_for_wkw(
-                            slices, num_channels=tile_size[2]
+    # Iterate over the z batches
+    # Batching is useful to utilize IO more efficiently
+    for z_batch in get_chunks(z_batches, batch_size):
+        try:
+            ref_time = time.time()
+            logging.info("Cubing z={}-{}".format(z_batch[0], z_batch[-1]))
+
+            for x in range(min_dimensions["x"], max_dimensions["x"] + 1):
+                for y in range(min_dimensions["y"], max_dimensions["y"] + 1):
+                    ref_time2 = time.time()
+                    # Allocate a large buffer for all images in this batch
+                    # Shape will be (channel_count, x, y, z)
+                    # Using fortran order for the buffer, prevents that the data has to be copied in rust
+                    buffer_shape = [
+                        num_channels,
+                        tile_size[0],
+                        tile_size[1],
+                        batch_size,
+                    ]
+                    buffer = np.empty(buffer_shape, dtype=dtype, order="F")
+                    for z in z_batch:
+                        # Read file if exists or use zeros instead
+                        file_name = find_file_with_dimensions(
+                            input_path_pattern, x, y, z, decimal_lengths
                         )
-
-                        if np.any(buffer != 0):
-                            target_wkw.write(
-                                [x * tile_size[0], y * tile_size[1], z_batch[0]], buffer
+                        if file_name:
+                            # read the image
+                            image = read_image_file(
+                                file_name,
+                                target_view.header.voxel_type,
+                                z,
+                                None,
+                                None,
                             )
-                        logging.debug(
-                            "Cubing of z={}-{} x={} y={} took {:.8f}s".format(
-                                z_batch[0], z_batch[-1], x, y, time.time() - ref_time2
+                        else:
+                            # add zeros instead
+                            image = np.zeros(
+                                tile_size + (1,),
+                                dtype=target_view.header.voxel_type,
                             )
+                        buffer[:, :, :, z - z_batch[0]] = image.transpose((2, 0, 1, 3))[
+                            :, :, :, 0
+                        ]
+
+                    if np.any(buffer != 0):
+                        target_view.write(data=buffer)
```

Comment:
I think, this line is missing the offset which is constituted by `[x * tile_size[0], y * tile_size[1]`. Otherwise, all tiles would be written to `[0, 0]`.